### PR TITLE
Move multidata to dev + fix bugs

### PR DESCRIPTION
**Pull request by felipemello1**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
- moved recipe to dev
- Updated checkpointing, to make it work

ISSUES: The dataloader does NOT resume from ckpt. This means that the order of the shuffle for a new epoch will be different vs checkpoint. This is NOT a blocking issue. When moving to step based checkpoint, we can resume the dataloader from ckpt. This is currently not possible because `loader` does not have method "set_epoch". We could go around it by adding StatefulDataLoader, but this is beyond the scope of this PR.

#### Test plan

```
tune run --nproc_per_node 2 dev/lora_finetune_distributed_multi_dataset --config dev/11B_lora_multi_dataset epochs=2 max_steps_per_epoch=50 metric_logger=torchtune.training.metric_logging.WandBLogger
```

image

### CODE REVIEW

File: recipes/dev/lora_finetune_distributed_multi_dataset.py [hunk 1]
```diff
@@ -489,11 +489,9 @@ def _setup_model(
                 ) and not lora_weights_state_dict:
                     # lora may not be covered in state dict
                     # if finetune for the 1st time
-                    m.lora_a.to_empty(device=lora_device)
-                    m.lora_b.to_empty(device=lora_device)
                     m.to_empty(device=lora_device)
                     m.initialize_parameters()
-                # RoPE is not covered in state dict
```
Comment (felipemello1) [inline recipes/dev/lora_finetune_distributed_multi_dataset.py:496]: copied from full_finetune_distributed

File: recipes/dev/lora_finetune_distributed_multi_dataset.py [hunk 2]
```diff
@@ -689,14 +689,11 @@ def save_checkpoint(
...
         # To prevent GPU memory from spiking during checkpoint save,
         # we consolidate the full model and optim state dicts on CPU for rank 0
-        state_dict = self._model.state_dict()
-        if self._save_adapter_weights_only:
-            state_dict = get_adapter_state_dict(state_dict, device=None)
-
         cpu_state_dict = training.gather_cpu_state_dict(
-            state_dict,
+            self._model,
             self._is_rank_zero,
```
Comment (felipemello1) [inline recipes/dev/lora_finetune_distributed_multi_dataset.py:694]: copied from full_finetune_distributed

File: torchtune/_recipe_registry.py [hunk 3]
```diff
@@ -438,12 +438,12 @@ class Recipe:
...
-        name="lora_finetune_distributed_multi_dataset",
-        file_path="lora_finetune_distributed_multi_dataset.py",
+        name="dev/lora_finetune_distributed_multi_dataset",
+        file_path="dev/lora_finetune_distributed_multi_dataset.py",
         configs=[
             Config(
-                name="llama3_2_vision/11B_lora_multi_dataset",
-                file_path="llama3_2_vision/11B_lora_multi_dataset.yaml",
+                name="dev/11B_lora_multi_dataset",
+                file_path="dev/11B_lora_multi_dataset.yaml",
```
Comment (ebsmothers) [inline torchtune/_recipe_registry.py:446]: nit: it might be worth keeping some version of `llama3_2_vision` in the filename (otherwise it's not immediately clear what this config is)

File: recipes/dev/lora_finetune_distributed_multi_dataset.py [hunk 4]
```diff
@@ -314,6 +314,11 @@ def setup(self, cfg: DictConfig) -> None:
             cfg_dataloader=cfg.dataloader,
             cfg_datasets=cfg.datasets,
             batch_size=cfg.batch_size,
+            dataloader_state_dict=(
+                checkpoint_dict[training.DATALOADER_KEY]
+                if self._resume_from_checkpoint
+                else None
+            ),
```
Comment (ebsmothers) [inline recipes/dev/lora_finetune_distributed_multi_dataset.py:321]: Shouldn't this be removed?

**Review by ebsmothers — APPROVED** [review]
A couple small comments, otherwise looks good

File: torchtune/_recipe_registry.py (hunk 3 above)
Comment (felipemello1) [reply torchtune/_recipe_registry.py:446]: i agree with the logic, but its not the pattern we use for any configs in /dev. Change it still?

File: recipes/dev/lora_finetune_distributed_multi_dataset.py (hunk 4 above)
Comment (felipemello1) [reply recipes/dev/lora_finetune_distributed_multi_dataset.py:321]: i thought about it. Decided to leave it as a placeholder for when we merge step based ckpt, but it can be deceiving that we save the state dict but dont use it. It is just a small dict though.

I can remove if you think its better

Comment (ebsmothers) [reply recipes/dev/lora_finetune_distributed_multi_dataset.py:321]: Personally I would remove, but it's not a blocker

File: torchtune/_recipe_registry.py (hunk 3 above)
Comment (ebsmothers) [reply torchtune/_recipe_registry.py:446]: OK then it's fine to leave as is
